src/Ryujinx.Graphics.Vulkan/BarrierBatch.cs

```diff
@@ -0,0 +1,228 @@
+                int bufferCount = 0;
+                int imageCount = 0;
+
+                bool hasBarrier = false;
+                StageFlags flags = default;
+
+                static void AddBarriers<T>(
+                    Span<T> target,
+                    ref int queuedBarrierCount,
+                    ref bool hasBarrier,
+                    ref StageFlags flags,
+                    ref int count,
+                    List<BarrierWithStageFlags<T>> list) where T : unmanaged
+                {
+                    int firstMatch = -1;
+
+                    for (int i = 0; i < list.Count; i++)
+                    {
+                        BarrierWithStageFlags<T> barrier = list[i];
+
+                        if (!hasBarrier)
+                        {
+                            flags = barrier.Flags;
+                            hasBarrier = true;
+
+                            target[count++] = barrier.Barrier;
+                            queuedBarrierCount--;
+                            firstMatch = i;
+
+                            if (count >= target.Length)
+                            {
+                                break;
+                            }
+                        }
+                        else
+                        {
+                            if (flags.Equals(barrier.Flags))
+                            {
+                                target[count++] = barrier.Barrier;
+                                queuedBarrierCount--;
+
+                                if (firstMatch == -1)
+                                {
+                                    firstMatch = i;
+                                }
+
+                                if (count >= target.Length)
+                                {
+                                    break;
+                                }
+                            }
+                            else
+                            {
+                                // Delete consumed barriers from the first match to the current non-match.
+                                if (firstMatch != -1)
+                                {
+                                    int deleteCount = i - firstMatch;
+                                    list.RemoveRange(firstMatch, deleteCount);
+                                    i -= deleteCount;
+
+                                    firstMatch = -1;
+                                }
+                            }
+                        }
+                    }
+
+                    if (firstMatch != -1)
+                    {
+                        if (firstMatch == 0)
+                        {
+                            list.Clear();
+                        }
+                        else
+                        {
+                            int deleteCount = list.Count - firstMatch;
+
+                            list.RemoveRange(firstMatch, deleteCount);
+                        }
+                    }
```

Comment:
Probably easier to read:
```suggestion
                    if (firstMatch == 0)
                    {
                        list.Clear();
                    }
                    else if (firstMatch != -1)
                    {
                        int deleteCount = list.Count - firstMatch;

                        list.RemoveRange(firstMatch, deleteCount);
                    }
```

Reply:
I'm not sure, there's more conditions but the wrapping `firstMatch != -1` means that "there is a match", and whether to clear or trim the list comes from that.

Reply:
I'll go for it, since
```
                    if (firstMatch != -1)
                    {
                        if (firstMatch == 0)
                        {
```
does look kind of bad.